### ISSUE TITLE
DONE / NEGATIVE EXP-002: generic note-graph topology did not beat flat same-feature baseline

### ISSUE THREAD

**Issue by gr-rr** [ListenCloser/listencloser#1305]
Program owner: #1303.
Depends conceptually on #1304 for reusable frozen symbolic features, but may proceed independently with simple note features if that keeps the ablation cleaner.
Research synthesis: #1297.
Relevant product/research owners: #1248 (AnalysisGNN), #1178 (symbolic detail), #1300 (bespoke cleanup).

## Question

> **Does explicit musical graph topology add held-out predictive value beyond the exact same node features?**

This is the smallest rigorous test of the broader “encode music as graphs” thesis. It is **not** permission to build a graph database, new persistence model, or generic graph platform.

## Why this is worth testing

GraphMuse/MusGConv/AnalysisGNN suggest that note relations such as succession, simultaneity, overlap, and metrical hierarchy can improve symbolic analysis. ListenCloser already has exact note, beat, measure, Score, and alignment identity, so a graph view is a plausible ML substrate.

But the graph hypothesis should earn itself through an ablation before architecture grows around it.

## OSS starting point

Prefer maintained existing components:
- GraphMuse: https://github.com/manoskary/graphmuse
- GraphMuse paper: https://arxiv.org/abs/2407.12671
- MusGConv: https://arxiv.org/abs/2405.09224
- PyTorch Geometric
- Partitura for symbolic parsing when needed

Record exact code/license versions. Do not fork/reimplement graph sampling/message passing unless the upstream path cannot support the bounded experiment.

## Experimental design

Choose **one primary note-level task** with legitimate labels and a clear LC motivation. Preferred task: motif-note importance/membership if a clean dataset is available. Acceptable alternates include voice/texture-related note labeling, cadence-related note/context labeling, or another symbolic task where graph relations are plausibly causal and labels are established.

Freeze the task and split before the result-bearing comparison.

### Node features

Use the exact same node features for both models. Example bounded feature set:

- pitch / pitch class;
- duration;
- metrical position;
- velocity when semantically appropriate;
- optional frozen #1304 representation embedding, if available.

Do not give the graph model extra semantic features unavailable to the MLP.

### Baseline

```text
node features
→ small MLP
→ task label
```

### Graph candidate

```text
same node features
+ graph topology
→ small GraphMuse/PyG GNN
→ same task label
```

Initial edge/hierarchy set should stay small and musically literal:

- follows / adjacency;
- simultaneous-onset;
- overlap/sustain relation;
- optional beat membership;
- optional measure membership.

Do not introduce learned/heuristic “motif”, “voice”, “harmony” edges that leak the target.

## Required ablations

At minimum compare:

```text
A. MLP, no topology
B. GNN, literal note-note topology
C. GNN + metrical hierarchy only if B is promising
```

Optional: random-edge or shuffled-edge negative control if implementation is cheap. This can test whether gains come from meaningful topology rather than parameter count.

## Metrics

Primary: task-native held-out F1/accuracy.

Also report:
- parameter count;
- training/inference wall time;
- peak memory if easy to measure;
- mean +/- variation across 3 seeds when training variance matters.

Primary decision:

> **Does the graph model beat the same-feature MLP by a meaningful, repeatable held-out margin?**

Use bootstrap confidence intervals or seed variation when practical; do not over-interpret a tiny single-run delta.

## Compute budget

```text
GPU-hours <= 4
cash cost <= $5
no broad search over architectures
```

Preferred runtime:
- local CPU for graph construction/data validation;
- free Kaggle CUDA for PyG/GraphMuse training if local/MPS support is awkward;
- Modal/RunPod only for a bounded reproducible rerun.

A tiny GNN should fit comfortably in this envelope. If it does not, stop and document why.

## Deliverables

1. Fixed task/dataset/split manifest.
2. Deterministic graph-construction function with explicit literal edge semantics.
3. Same-feature MLP baseline.
4. Small GraphMuse/PyG candidate.
5. Result-bearing table including quality, params, runtime, seeds.
6. Error analysis on cases where topology helps/hurts.
7. One disposition:

```text
FAIL — graph topology adds no meaningful value
NARROW — helps one task but no shared graph thesis yet
CONTINUE — meaningful lift justifies testing graph heads on another existing LC task
```

## Stop / kill criteria

Stop rather than escalating if:

- topology provides no repeatable held-out gain;
- gain disappears under same-feature/parameter-control comparison;
- graph construction requires target-leaking semantic heuristics;
- dependency/runtime complexity is disproportionate to the measured lift.

A negative result is useful and should argue against generic graph infrastructure.

## Non-goals

- no Neo4j/graph database;
- no production graph persistence;
- no generic “MusicGraph” service/framework;
- no large heterogeneous graph schema;
- no AnalysisGNN production replacement decision;
- no multiple-task expansion until the one-task ablation is conclusive;
- no model selection based on paper numbers alone.

## Acceptance

- [ ] one exact LC-relevant task and fixed split selected;
- [ ] MLP and GNN consume identical node features;
- [ ] graph edges are literal/non-target-leaking;
- [ ] held-out quality and compute compared reproducibly;
- [ ] <=4 GPU-hours / <=$5 unless stopped earlier with reason;
- [ ] explicit FAIL/NARROW/CONTINUE disposition written to #1303;
- [ ] no graph infrastructure is introduced outside the research boundary without separate evidence-backed ownership.

Refs #1303 #1304 #1297 #1248 #1178 #1300.

**Comment by gr-rr** [comment]
## Kickoff / clean-ablation refinement

Run after or alongside #1304 data preparation, reusing a public labeled note task where possible. **MNID is the preferred first task** if its exact processed split/rights are admitted by #1304.

### Primary experiment must isolate topology

Do not use contextual M2BERT embeddings in the primary A/B comparison. That would already inject neighborhood/context information into every node and make the graph question harder to interpret.

Freeze a small literal feature vector shared exactly by both models, e.g.:

```text
pitch / pitch class
duration
metrical position
optional velocity only if legitimate for the task
```

Then compare:

```text
A. same node features -> small MLP
B. same node features + literal graph edges -> small GNN
```

Only if B gives a repeatable held-out lift may an optional secondary condition add the frozen #1304 representation to both models.

### Topology v0

Keep edges mechanically derivable and non-target-leaking:
- follows/adjacency;
- simultaneous onset;
- overlap/sustain;
- beat/measure membership only if useful and exact.

Prefer an edge-shuffle/random-edge negative control if it is nearly free; it is more informative than another architecture variant.

### Compute

Graph construction/data checks: local CPU.
Tiny MLP: local CPU.
Tiny GNN: local CPU first; use #1311 Kaggle T4 only for Linux/PyG/GraphMuse dependency friction or clear wall-time benefit.

Record params, wall time, peak memory where practical, and 3 seeds. No architecture search.

### Decision

The result should answer only: **does literal topology add held-out value beyond identical node features?** A negative result should actively argue against generic graph infrastructure.

Refs #1303 #1304 #1311 #1248 #1300.

**Comment by gr-rr** [comment]
Execution kickoff: branch `research/1305-graph-ablation` created from current main (`d11f0c46...`). This experiment may proceed in parallel with #1304 and should **not** wait for M2BERT. Primary ablation should use literal/local node features first so the graph-vs-no-graph question stays clean; contextual M2BERT embeddings can be an optional follow-up only if topology shows value. Prefer GraphMuse/PyG/Partitura upstream APIs, no custom GNN framework. Use one existing labeled symbolic task, fixed source-disjoint split, same features for MLP/GNN, 3 seeds only if variance matters. Kaggle notebook T4 path is now proven; CPU graph construction + bounded T4 training is the preferred execution split.

**Comment by gr-rr** [comment]
Execution update — first result-bearing graph-vs-no-graph ablation completed on free GitHub CPU.

Fixed setup:
- task: BPS motif note identification, official CV fold 0 (train folds 2/3/4, valid 1, test 0)
- same static 154-d one-hot CP note features for every model
- literal topology only: bidirectional score-order adjacency + simultaneous (Bar, Position) onset edges
- candidate: standard `torch_geometric.nn.SAGEConv`; no custom message passing
- random-edge negative control included
- 3 seeds; ~19.6k train nodes, ~9.2k valid, ~19.2k test

Held-out test F1 (mean ± std across seeds):
- flat MLP: **0.6079 ± 0.0042**
- GraphSAGE + literal musical edges: **0.5996 ± 0.0056**
- GraphSAGE + random edges: **0.5969 ± 0.0064**

Parameter counts: flat MLP 14,210; GraphSAGE 28,162.

Interpretation: literal graph topology did **not** add predictive value on this clean same-feature motif-membership ablation; it was slightly worse than the flat MLP, and only marginally above the random-edge control. This satisfies the issue's stop criterion rather than justifying metrical hierarchy / broader graph infrastructure.

Disposition: **FAIL for the generic graph-topology thesis on this task.** Keep GraphMuse / AnalysisGNN as OSS task-specific candidates where separately justified, but do not build a generic ListenCloser graph substrate from this evidence. A second task should only be run if there is a specific product hypothesis where topology is plausibly causal, not to rescue the broad graph idea.

**Comment by gr-rr** [comment]
## Final disposition — FAIL for generic note-graph topology; stop

Result-bearing ablation is complete and satisfies the pre-registered stop rule.

Fixed comparison:
- task: BPS motif-note identification, official CV fold 0;
- same 154-d literal CP note features for every model;
- flat MLP vs PyG GraphSAGE;
- literal graph edges = bidirectional score-order adjacency + simultaneous onset;
- random-edge negative control;
- 3 seeds.

Held-out test F1:

| model | F1 mean ± std |
|---|---:|
| flat MLP | **0.6079 ± 0.0042** |
| GraphSAGE + literal musical edges | **0.5996 ± 0.0056** |
| GraphSAGE + random edges | **0.5969 ± 0.0064** |

The literal graph did not beat the same-feature flat baseline and was only marginally above randomized topology. There is no evidence here for a generic ListenCloser graph substrate, graph database, or homegrown GNN program.

What this **does not** mean:
- graphs are not universally useless;
- AnalysisGNN / GraphMuse should not be rejected when their official task-specific models solve a concrete problem;
- a future focused task where hierarchy/topology is plausibly causal is forbidden.

What it **does** mean:
- no second generic graph experiment should be launched merely to rescue this thesis;
- consume OSS graph models task-by-task;
- any future graph ablation needs a new concrete product residual gap and its own focused owner.

Disposition: **FAIL / completed negative result.** Write this into #1303's synthesis and close #1305.
